src/cs/Ssh.Tcp/Services/PortForwardingService.cs

```diff
@@ -150,6 +151,32 @@ public PortForwardingService(SshSession session)
 	/// </remarks>
 	public ForwardedPortsCollection RemoteForwardedPorts { get; } = new ForwardedPortsCollection();
 
+	/// <summary>
+	/// Event raised when an incoming or outgoing connection to a forwarded port is
+	/// about to be established.
+	/// </summary>
+	public event EventHandler<ForwardedPortConnectingEventArgs>? ForwardedPortConnecting;
+
+	internal async Task<Stream?> OnForwardedPortConnectingAsync(
+		int port,
+		bool isIncoming,
+		SshStream stream,
+		CancellationToken cancellation)
+	{
+		var e = new ForwardedPortConnectingEventArgs(port, isIncoming, stream, cancellation);
+		ForwardedPortConnecting?.Invoke(this, e);
```

Comment:
For most cases the event handler will be `null`, and to save on GC, we can check the handler first:
``` csharp
var handler = ForwardedPortConnecting;
if (handler == null)
{
    return stream;
}

var e = new ForwardedPortConnectingEventArgs(port, isIncoming, stream, cancellation);
handler(this, e);
...
```

Reply:
Updated.